### PR TITLE
[DDW-970] Prevent creation of new wallet during address verification

### PR DESCRIPTION
**Pull request by szymonmaslowski**
This PR
 - prevents creating a new wallet when paired incorrect one during address verification
 - refactors part of HW logic

Unfortunately, this PR requires full regression testing of hardware wallet cases

Diagram representing part of the logic covered by the refactoring:
image

## Todos

- [x] Wait for [PR 2860](https://github.com/input-output-hk/daedalus/pull/2860) to be merged and use HWDeviceStatus introduced there instead of creating new one here.

## Screenshots

image
image

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1646746432146099)
- [ ] Full regression test of all hardware wallet cases. **[Regression document](https://docs.google.com/spreadsheets/d/1fDAaHRohJ4qBrrRrpIJlYmp6h2AidFegbGMe5IU-bUg/edit?usp=sharing)**

---

### Basics

- [x] PR assigned to the PR author(s)
- [x] `input-output-hk/daedalus-dev` and `input-output-hk/daedalus-qa` assigned as PR reviewers
- [x] If there are UI changes, Alexander Rukin assigned as an additional reviewer
- [x] All visual regression testing has been reviewed (assign `run Chromatic` label to PR to trigger the run)
- [x] PR has appropriate labels (`release-vNext`, `feature`/`bug`/`chore`, `WIP`)
- [x] PR link is added to a Jira ticket, ticket moved to In Review
- [x] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [x] PR has a good description that summarizes all changes
- [ ] PR contains screenshots (in case of UI changes)
- [x] CHANGELOG entry has been added to the top of the appropriate section (_Features_, _Fixes_, _Chores_) and is linked to the correct PR on GitHub
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [x] In case of dependency changes `yarn.lock` file is updated

### Code Quality

- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with typescript types
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing

- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review

- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: source/renderer/app/components/hardware-wallet/HardwareWalletStatus.tsx
Comment (lucas-barros) [inline source/renderer/app/components/hardware-wallet/HardwareWalletStatus.tsx:242]: Remove console.logs

Comment (szymonmaslowski) [reply source/renderer/app/components/hardware-wallet/HardwareWalletStatus.tsx]: done already :)

**Review by lucas-barros — APPROVED** [review]

**Review by shawnbusuttil — APPROVED** [review]

**Comment by alexander-rukin** [comment]
in the bottom there is too much space (because there is no buttons, but paddings are there 😓)

image

20px here 

image

and 12px here

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (tomislavhoracek) [inline source/renderer/app/stores/HardwareWalletsStore.ts:729]: @szymonmaslowski This should remain `_getExtendedPublicKey` because that is the main procedure and we are identifying Hardware-Software just to check if we already have an extended public key that matches some software wallet

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (tomislavhoracek) [inline source/renderer/app/stores/HardwareWalletsStore.ts:781]: @szymonmaslowski same as above comment

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (tomislavhoracek) [inline source/renderer/app/stores/HardwareWalletsStore.ts:905]: @szymonmaslowski same as above comment

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (tomislavhoracek) [inline source/renderer/app/stores/HardwareWalletsStore.ts:1009]: @szymonmaslowski same as above comment

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (tomislavhoracek) [inline source/renderer/app/stores/HardwareWalletsStore.ts:1263]: @szymonmaslowski following to the above comment. Now you have 2 functions that are connected to identifying / recognizing. It is better to say "ok, now get me an extended public key and then recognize"

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (tomislavhoracek) [inline source/renderer/app/stores/HardwareWalletsStore.ts:1648]: @szymonmaslowski maybe we should call this `_reinitiate... or _recover...`? What do you think?

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (tomislavhoracek) [inline source/renderer/app/stores/HardwareWalletsStore.ts:1742]: We can remove this empty line

**Review by tomislavhoracek — CHANGES_REQUESTED** [review]
@szymonmaslowski 💯  Nice improvements and code / flow is much more cleaner now.
Please check my comments...

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (tomislavhoracek) [reply source/renderer/app/stores/HardwareWalletsStore.ts:729]: Ouu, I see now that you have dedicated fn for `_getExtendedPublicKey` and this is just for matching. 
Ignore the comment! Resolved!

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (szymonmaslowski) [reply source/renderer/app/stores/HardwareWalletsStore.ts:1648]: I would say this is rather a continuation of the transaction process as there is the `_signTransactionTrezor` method called at the end of this method. Am I right?

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (szymonmaslowski) [reply source/renderer/app/stores/HardwareWalletsStore.ts:1263]: Sorry I didn't get your comment. By two functions do you mean the `_handleIdentifiedSoftwareWallet` and `_identifyAndHandleSoftwareWallet`? The `_handleIdentifiedSoftwareWallet` method is responsible only for the case when the wallet was found. On the other hand the `_identifyAndHandleSoftwareWallet` method encapsulates all steps including the extended public key request, finding the device and handling it accordingly, and it is there for a reduction of code duplication, but I believe with further refactoring it may be broken to few less generic code blocks and distributed to the proper places in the logic (for example part regarding the transaction to the transaction logic etc.).

**Comment by szymonmaslowski** [comment]
@alexander-rukin Could you check again? Styling issues shouldn't be there anymore. I fixed it in a little different way than your suggestion, but I think I achieved the same result.

File: source/renderer/app/components/hardware-wallet/HardwareWalletStatus.tsx
Comment (szymonmaslowski) [reply source/renderer/app/components/hardware-wallet/HardwareWalletStatus.tsx]: Done :+1:

**Comment by alexander-rukin** [comment]
@szymonmaslowski pls make sure that your fix didn't affect scrollbar position 

image

as now it ends too early (not with the end of content)

**Comment by szymonmaslowski** [comment]
@alexander-rukin and I decided to create separate ticket for fixing UI issues of the address dialog because the issue(s) require some effort and they are not related to changes of this PR.

**Review by marcin-mazurek — APPROVED** [review]
Good job - solid refactoring :) I'm only concerned with the function name: `_handleAssociatedWallet `, if you can try to turn it into a more descriptive name (even if it's very long), it would be fantastic. Everything else was discussed during the group review session and already addressed.

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (marcin-mazurek) [inline source/renderer/app/stores/HardwareWalletsStore.ts:1651]: In this case, should we not also assign the `INCORRECT_PASSPHRASE_PROVIDED` status? (or like we discussed - UNRECOGNIZED_WALLET). If you check the implementation of the method which calls this method - the implementation and the comment suggest it is the same case:

```
if (!expectedWalletId || associatedWallet.id !== expectedWalletId) {
        logger.debug('[HW-DEBUG] HWStore - Device not belongs to this wallet');
        this._discardConnectedDeviceAndReInitiateTransaction({
          walletId: expectedWalletId,
        });
```

**Review by marcin-mazurek — CHANGES_REQUESTED** [review]

File: source/renderer/app/stores/HardwareWalletsStore.ts
Comment (szymonmaslowski) [reply source/renderer/app/stores/HardwareWalletsStore.ts:1651]: That makes sense. The original issue was not connected to the scenario of the transaction so I didn't touch this case and during the refactoring just copied over the implementation. I will try to change the status to the one you suggested and reproduce that case to check it works fine :+1:

Comment (marcin-mazurek) [reply source/renderer/app/stores/HardwareWalletsStore.ts:1651]: That is also covered by my PR: https://github.com/input-output-hk/daedalus/pull/2860
I just noticed that because if we were to merge your refactoring as-is, my change would be reverted (assuming it will be merged first)

**Comment by gabriela-ponce** [comment]
@szymonmaslowski I found a specific scenario where the message is different from the one on the screenshots:

- Pair the Trezor and do not unplug it. Then, try to validate the address and insert an incorrect recovery phrase. You'll see no new device is paired, but the error will be "Address verification failed" instead of "We do not recognize this wallet...". Check [this video](https://drive.google.com/file/d/1qKtcsOIS5EpsLzs6fKe3ppxeFI1ydQS5/view?usp=sharing)
Could this be improved and show the second message instead?

**Review by gabriela-ponce — APPROVED** [review]
Good job, lgtm!
Notes: 
- The issue on [my previous comment](https://github.com/input-output-hk/daedalus/pull/2906#issuecomment-1111345573) is going to be addressed on this card: [DDW-1090](https://input-output.atlassian.net/browse/DDW-1090).
- Please do not merge until we have at least one additional approval from the rest of @input-output-hk/daedalus-qa

**Review by miorsufianiohk — APPROVED** [review]
LGTM. Great job @szymonmaslowski 👍 . Tested on 21656

**Comment by dmitrii-gaico** [comment]
@danielmain Additional QA review is required after resolving the conflicts. So please do not merge before further notice.

**Review by ManusMcCole — APPROVED** [review]
Looks good @szymonmaslowski 👍

**Review by gabriela-ponce — APPROVED** [review]

**Review by miorsufianiohk — APPROVED** [review]
LGTM. Tested on 21788. Great work @szymonmaslowski 👍

**Review by danielmain — APPROVED** [review]
Amazing job @szymonmaslowski

**Pull request merged**
